protocol/shannon/websocket_context.go

```diff
@@ -0,0 +1,184 @@
+	if sessionHeader == nil {
+		rc.logger.Error().Msg("❌ SHOULD NEVER HAPPEN: Error getting relay miner connection headers: session header is nil")
+		return http.Header{}
+	}
+
+	return http.Header{
+		request.HTTPHeaderTargetServiceID: {sessionHeader.ServiceId},
+		request.HTTPHeaderAppAddress:      {sessionHeader.ApplicationAddress},
+		proxy.RPCTypeHeader:               {strconv.Itoa(int(sharedtypes.RPCType_WEBSOCKET))},
+	}
+}
+
+// GetWebsocketEndpointURL returns the websocket URL for the selected endpoint.
+// This URL is used to establish the websocket connection to the endpoint.
+func (rc *requestContext) GetWebsocketEndpointURL() (string, error) {
+	websocketURL, err := rc.selectedEndpoint.WebsocketURL()
+	if err != nil {
+		rc.logger.Error().Err(err).Msg("❌ Selected endpoint does not support websocket RPC type")
+		return "", err
+	}
+
+	return websocketURL, nil
+}
+
+func (rc *requestContext) SignClientWebsocketMessage(msgData []byte) ([]byte, error) {
+	unsignedRelayRequest := &servicetypes.RelayRequest{
+		Meta: servicetypes.RelayRequestMetadata{
+			SessionHeader:           rc.selectedEndpoint.Session().GetHeader(),
+			SupplierOperatorAddress: rc.selectedEndpoint.Supplier(),
+		},
+		Payload: msgData,
+	}
+
+	app := rc.selectedEndpoint.Session().GetApplication()
+	if app == nil {
+		rc.logger.Error().Msg("❌ SHOULD NEVER HAPPEN: session application is nil")
+		return nil, fmt.Errorf("session application is nil")
+	}
+
+	signedRelayRequest, err := rc.relayRequestSigner.SignRelayRequest(unsignedRelayRequest, *app)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", errRelayRequestSigningFailed, err.Error())
+	}
+
+	relayRequestBz, err := signedRelayRequest.Marshal()
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", errRelayRequestSigningFailed, err.Error())
+	}
+
+	return relayRequestBz, nil
+}
+
+func (rc *requestContext) ValidateEndpointWebsocketMessage(msgData []byte) ([]byte, error) {
+	// Validate the relay response using the Shannon FullNode
+	relayResponse, err := rc.fullNode.ValidateRelayResponse(sdk.SupplierAddress(rc.selectedEndpoint.Supplier()), msgData)
+	if err != nil {
+		rc.logger.Error().Err(err).Msg("❌ failed to validate relay response")
+		return nil, fmt.Errorf("%w: %s", errRelayResponseValidationFailed, err.Error())
+	}
+
+	rc.logger.Debug().Msgf("received message from protocol endpoint: %s", string(relayResponse.Payload))
+
+	return relayResponse.Payload, nil
+}
+
+// TODO_IN_THIS_PR(@commoddity): clean up the observation initialization logic below
+// TODO_IN_THIS_PR(@commoddity): create a new Shannon observation specific to websockets,
+// which must differentiate between a failed connection attempt and a failed message.
+
+// UpdateMessageObservationsFromSuccess updates the observations for the current message
+// if the message handler does not return an error.
+func (rc *requestContext) UpdateMessageObservationsFromSuccess(
```

Comment:
The protocol/shannon package should decide if a message from client/endpoint was a success or failure, and update its own observation (to provide when requested using e.g. `GetEndpointObservation()`).
This should not be a public method (placing the burden on the caller to determine success/failure from a protocol perspective).